app/ui/legacy/src/main/java/com/fsck/k9/ui/account/AccountsViewModel.kt

```diff
@@ -1,8 +1,118 @@
+import com.fsck.k9.controller.UnreadMessageCountProvider
+import com.fsck.k9.provider.EmailProvider
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.channels.awaitClose
+import kotlinx.coroutines.channels.sendBlocking
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.callbackFlow
+import kotlinx.coroutines.flow.collect
+import kotlinx.coroutines.flow.combine
+import kotlinx.coroutines.flow.flatMapLatest
+import kotlinx.coroutines.launch
+import timber.log.Timber
 
-class AccountsViewModel(preferences: Preferences) : ViewModel() {
-    val accountsLiveData = AccountsLiveData(preferences)
+@ExperimentalCoroutinesApi
+class AccountsViewModel(
+    val preferences: Preferences,
+    val unreadMessageCountProvider: UnreadMessageCountProvider,
+    private val contentResolver: ContentResolver
+) :
+    ViewModel() {
+
+    private val mutableAccountAndUnreadsLiveData = MutableLiveData<List<DisplayAccount>>()
+
+    val displayAccountsLiveData: LiveData<List<DisplayAccount>>
+        get() = mutableAccountAndUnreadsLiveData
+
+    private val accountsFlow: Flow<List<Account>> =
+        callbackFlow {
+            Timber.d("getAccountsFlow.callbackFlow")
+
+            send(preferences.accounts)
+
+            Timber.d("sent.accounts")
+
+            val accountsChangeListener = AccountsChangeListener {
+                Timber.d("onAccountsChanged")
+                sendBlocking(preferences.accounts)
+            }
+            preferences.addOnAccountsChangeListener(accountsChangeListener)
+            awaitClose {
+                Timber.d("removeOnAccountsChangeListener")
+                preferences.removeOnAccountsChangeListener(accountsChangeListener)
+            }
+        }
+
+    private val displayAccountFlow: Flow<List<DisplayAccount>> = accountsFlow
+        .flatMapLatest { accounts ->
+            Timber.d("displayAccountFlow.flatMapLatest; accounts: %s", accounts)
+
+            val unreadCountFlows: List<Flow<Int>> = accounts.map { account ->
+                Timber.d("account: %s", account.uuid)
+                getUnreadCountFlow(account)
+            }
+
+            combine(unreadCountFlows) { unreadCounts ->
+                Timber.d("unreadCounts: %s", unreadCounts)
+                unreadCounts.mapIndexed { index, unreadCount ->
+                    Timber.d("index, unreadCount: %d, %s", index, unreadCounts)
+                    DisplayAccount(account = accounts[index], unreadCount)
+                }
+            }
+        }
+
+    private fun getUnreadCountFlow(account: Account): Flow<Int> {
+        return callbackFlow {
+            Timber.d("getUnreadCountFlow.callbackFlow; account: %s", account)
+
+            val notificationUri = EmailProvider.getNotificationUri(account.uuid)
+
+            viewModelScope.launch(Dispatchers.IO) {
+                send(unreadMessageCountProvider.getUnreadMessageCount(account))
+            }
```

Comment:
By using the `flowOn` below can simplify this:

```suggestion
            send(unreadMessageCountProvider.getUnreadMessageCount(account))
```